packages/block-serialization-default-parser/parser.php

```diff
@@ -0,0 +1,449 @@
+
+/**
+ * Class WP_Block_Parser_Block
+ *
+ * Holds the block structure in memory
+ *
+ * @since 3.8.0
+ */
+class WP_Block_Parser_Block {
+	/**
+	 * Name of block
+	 *
+	 * @example "core/paragraph"
+	 *
+	 * @since 3.8.0
+	 * @var string
+	 */
+    public $blockName;
+
+	/**
+	 * Optional set of attributes from block comment delimiters
+	 *
+	 * @example null
+	 * @example array( 'columns' => 3 )
+	 *
+	 * @since 3.8.0
+	 * @var array|null
+	 */
+    public $attrs;
+
+	/**
+	 * List of inner blocks (of this same class)
+	 *
+	 * @since 3.8.0
+	 * @var WP_Block_Parser_Block[]
+	 */
+    public $innerBlocks;
+
+	/**
+	 * Resultant HTML from inside block comment delimieters
+	 * after removing inner blocks
+	 *
+	 * @example "...Just <!-- wp:test /--> testing..." -> "Just testing..."
+	 *
+	 * @since 3.8.0
+	 * @var string
+	 */
+    public $innerHTML;
+
+    function __construct( $name, $attrs, $innerBlocks, $innerHTML ) {
+        $this->blockName   = $name;
+        $this->attrs       = $attrs;
+        $this->innerBlocks = $innerBlocks;
+        $this->innerHTML   = $innerHTML;
+    }
+}
+
+/**
+ * Class WP_Block_Parser_Frame
+ *
+ * Holds partial blocks in memory while parsing
+ *
+ * @internal
+ * @since 3.8.0
+ */
+class WP_Block_Parser_Frame {
+	/**
+	 * Full or partial block
+	 *
+	 * @since 3.8.0
+	 * @var WP_Block_Parser_Block
+	 */
+    public $block;
+
+	/**
+	 * Byte offset into document for start of parse token
+	 *
+	 * @since 3.8.0
+	 * @var int
+	 */
+    public $token_start;
+
+	/**
+	 * Byte length of entire parse token string
+	 *
+	 * @since 3.8.0
+	 * @var int
+	 */
+    public $token_length;
+
+	/**
+	 * Byte offset into document for after parse token ends
+	 * (used during reconstruction of stack into parse production)
+	 *
+	 * @since 3.8.0
+	 * @var int
+	 */
+    public $prev_offset;
+
+	/**
+	 * Byte offset into document where leading HTML before token starts
+	 *
+	 * @since 3.8.0
+	 * @var int
+	 */
+    public $leading_html_start;
+
+    function __construct( $block, $token_start, $token_length, $prev_offset = null, $leading_html_start = null ) {
+        $this->block = $block;
+        $this->token_start = $token_start;
+        $this->token_length = $token_length;
+        $this->prev_offset = isset($prev_offset) ? $prev_offset : $token_start + $token_length;
+        $this->leading_html_start = $leading_html_start;
+    }
+}
+
+/**
+ * Class WP_Block_Parser
+ *
+ * Parses a document and constructs a list of parsed block objects
+ *
+ * @since 3.8.0
+ */
+class WP_Block_Parser {
+	/**
+	 * Input document being parsed
+	 *
+	 * @example "Pre-text\n<!-- wp:paragraph -->This is inside a block!<!-- /wp:paragraph -->"
+	 *
+	 * @since 3.8.0
+	 * @var string
+	 */
+    public $document;
+
+	/**
+	 * Tracks parsing progress through document
+	 *
+	 * @since 3.8.0
+	 * @var int
+	 */
+    public $offset;
+
+	/**
+	 * List of parsed blocks
+	 *
+	 * @since 3.8.0
+	 * @var WP_Block_Parser_Block[]
+	 */
+    public $output;
```

Comment:
I'm concerned about this promise that `$output` is an array of `WP_Block_Parser_Block`, since freeform fragments are added as [associative] arrays and not class instances.

Reply:
we can definitely consider wiping the output clean of its classes - I didn't at first because it seemed benign to retain them, but if we sacrifice a little performance we can `json_decode( json_encode( $output ) )` and clear it up

Reply:
@jorgefilipecosta mentioned implementing an `ArrayObject` interface in our classes so that one can traverse our parser output natively, rather than doing the JSON dance. What do you think?

Reply:
That's a good question. It means more divide between the PHP and JS versions of the parser. What's the JSON dance? Wouldn't having `ArrayObject` be somewhat superfluous?

```php
// this already works with arrays and objects!
$blocks = parse( $document );
$blocks = array_map( $blocks, $my_transformer );
```

we probably want to fix the bug as a separate thing from adding interfaces. I'm skeptical of the value of the latter if the former is resolved.

Reply:
By JSON dance I meant `json_decode( json_encode( $output ) )`, sorry for not being clear.

> we probably want to fix the bug as a separate thing from adding interfaces

So this is the actual issue: #10047. It's not the traversal (looking at your `array_map`) example but rather accessing properties of a block, which can either mean accessing properties of an array or of an object.

Reply:
Classes offer some advantages we can publish abstract classes that contain the fields plugins can safely access, and other parsers can just extend this general classes. Simple arrays don't offer this guarantees.

But now we have a problem some plugins are dependent on using simple arrays, even if this bug was already caught I'm not certain we can just change the API to use classes.

So I think our options are revert back and just use arrays, or advance and change our API to use classes, but in this case in order to be back-compatible with existing implementations accessing using the array syntax I think our only solution is ArrayObject which allows us to temporarily return something that behaves like a class for new implementations and an array for old implementations, in this case, we would add the deprecation messages saying we now return objects.

Reply:
> It's not the traversal (looking at your array_map example) but rather accessing properties of a block, which can either mean accessing properties of an array or of an object.

to me this is just evidence that the work to make all attribute reporting consistent is necessary. some attributes are `null`, some are objects

> By JSON dance I meant json_decode( json_encode( $output ) ), sorry for not being clear.

that would be in the parser and wouldn't have to be manually performed. in fact, the classes are only even there for performance, so we can test the change of sorting everything in plain old objects vs. converting at the end. if it's a degradation then we can simply remove the classes if we want to preserve the simpler interface.